client/asset/btc/btc.go

```diff
@@ -1747,336 +1741,204 @@ func (btc *ExchangeWallet) LocktimeExpired(contract dex.Bytes) (bool, time.Time,
-				pendingContracts = append(pendingContracts, contractOutpoint)
+		for outPt, req := range undiscovered {
+			if iHeight >= req.blockHeight && reqStillQueued(req.outPt) {
+				validReqs[outPt] = req
 			}
 		}
 		btc.findRedemptionMtx.RUnlock()
-		btc.findRedemptionsInMempool(pendingContracts)
-	}
-}
 
-// findRedemptionsInTx checks if any input of the passed tx spends any of the
-// specified contract outpoints. If spending info is found for any contract, the
-// contract's secret or any error encountered while trying to parse the secret
-// is returned to the redemption finder via the registered result chan; and the
-// contract is purged from the findRedemptionQueue.
-// Returns the number of redemptions found and canceled.
-func (btc *ExchangeWallet) findRedemptionsInTx(scanPoint string, tx *btcjson.TxRawResult, contractOutpoints []outPoint) (found, cancelled int) {
-	btc.findRedemptionMtx.Lock()
-	defer btc.findRedemptionMtx.Unlock()
-
-	extractSecret := func(vin int, contractHash []byte) (*chainhash.Hash, []byte, error) {
-		redeemTxHash, err := chainhash.NewHashFromStr(tx.Txid)
-		if err != nil {
-			return nil, nil, err
-		}
-		if tx.Vin[vin].ScriptSig == nil {
-			return nil, nil, fmt.Errorf("no sigScript")
-		}
-		sigScript, err := hex.DecodeString(tx.Vin[vin].ScriptSig.Hex)
-		if err != nil {
-			return nil, nil, err
+		if len(validReqs) == 0 {
+			continue
 		}
-		var witness [][]byte
-		for _, hexB := range tx.Vin[vin].Witness {
-			var b []byte
-			b, err = hex.DecodeString(hexB)
-			if err != nil {
-				break
+
+		discovered := btc.node.searchBlockForRedemptions(validReqs, *iHash)
+		for outPt, res := range discovered {
+			req, found := undiscovered[outPt]
+			if !found {
+				btc.log.Errorf("Request not found in undiscovered map. This shouldn't be possible.")
+				continue
 			}
-			witness = append(witness, b)
-		}
-		txIn := wire.NewTxIn(new(wire.OutPoint), sigScript, witness)
-		secret, err := dexbtc.FindKeyPush(txIn, contractHash, btc.segwit, btc.chainParams)
-		if err != nil {
-			return nil, nil, err
+			req.success(res)
+			delete(undiscovered, outPt)
 		}
-		return redeemTxHash, secret, nil
-	}
 
-	for _, contractOutpoint := range contractOutpoints {
-		req, exists := btc.findRedemptionQueue[contractOutpoint]
-		if !exists {
-			continue // no find request for this outpoint (impossible now?)
-		}
-		if req.canceled() {
-			cancelled++
-			delete(btc.findRedemptionQueue, contractOutpoint)
-			continue // this find request has been cancelled
+		if len(undiscovered) == 0 {
+			return
 		}
 
-		for i := range tx.Vin {
-			input := &tx.Vin[i]
-			if input.Vout != contractOutpoint.vout || input.Txid != contractOutpoint.txHash.String() {
-				continue // input doesn't redeem this contract, check next input
+		iHeight += 1
+		if iHash, err = btc.node.getBlockHash(int64(iHeight)); err != nil {
+			if iHeight < tipHeight {
+				btc.log.Errorf("error getting block hash for height %d: %v", iHeight, err)
 			}
```

Comment:
Could also check here that iHash was equal to this new block's previous hash so we know we aren't in the middle of a reorg and can just stop. Or is `checkForNewBlocks()` canceling this remotely if there's a reorg, and leaving it to that? This function would continue however.